File: optimizers.py

class Optimizer:
    ''' Update weights and bias using optmizer '''

    @staticmethod
    def SGD(*layers, learning_rate=0.01):
        ''' Stochastic Gradient Descent optimizer '''

        for layer in layers:
            for k in range(layer.output_neurons):
                for j in range(layer.input_neurons):
                    layer.weights[k][j] -= learning_rate * layer.gradients[k][j]
                layer.bias[k] -= learning_rate * layer.bias_gradients[k]

File: nn/optimizers.c
#include "layers.h"

void SGD_layer(Layer* layer, float learning_rate) {
    for (size_t index = 0; index < layer->parameter_count; index++) {
        layer->weights[index] -= learning_rate * layer->gradients[index];
    }
    for (int output = 0; output < layer->output_neurons; output++) {
        layer->bias[output] -= learning_rate * layer->bias_gradients[output];
    }
}
